talk/owt/sdk/p2p/p2ppeerconnectionchannel.cc

```diff
@@ -163,6 +163,121 @@ void P2PPeerConnectionChannel::ClearPendingStreams() {
+      on_failure(std::move(e));
+    }
+    return;
+  }
+
+  auto stream_labels = std::vector<std::string>();
+  std::for_each(streams.begin(), streams.end(), [&stream_labels](std::shared_ptr<LocalStream> stream) {
+    if (stream) {
+      // Calling [stream|track]->id() runs on signaling_thread, so call outside of locks
+      stream_labels.push_back(stream->MediaStream()->id());
+    } else {
+      RTC_LOG(LS_INFO) << "Local stream cannot be nullptr. Skipping.";
+    }
+  });
+
+  {
+    std::lock_guard<std::mutex> lock(published_streams_mutex_);
+    for (const auto &stream_label : stream_labels) {
+      if (published_streams_.find(stream_label) != published_streams_.end() ||
+          publishing_streams_.find(stream_label) != publishing_streams_.end()) {
+        if (on_failure) {
+          std::unique_ptr<Exception> e(
+              new Exception(ExceptionType::kP2PClientInvalidArgument,
+                            "The stream is already published."));
+          on_failure(std::move(e));
+        }
+      }
+    }
+  }
+
+  // Prune null streams
+  streams.erase(std::remove_if(streams.begin(), streams.end(), [](std::shared_ptr<LocalStream> stream) {
+    return !stream;
+  }), streams.end());
+
+
+  for (auto stream : streams) {
+    RTC_LOG(LS_INFO) << "Publishing a local stream.";
+
+    // Send chat-closed to workaround known browser bugs, together with
+    // user agent information once and once only.
+    {
+      std::lock_guard<std::mutex> lock(stop_send_mutex_);
+      if (stop_send_needed_) {
+        SendStop(nullptr, nullptr);
+        stop_send_needed_ = false;
+      }
+      if (!ua_sent_) {
+        SendUaInfo();
+        ua_sent_ = true;
+      }
+    }
+    rtc::scoped_refptr<webrtc::MediaStreamInterface> media_stream =
+        stream->MediaStream();
+    // Calling [stream|track]->id() runs on signaling_thread, so call outside of locks
+    std::string stream_id = media_stream->id();
+    std::pair<std::string, std::string> stream_track_info;
+    for (const auto& track : media_stream->GetAudioTracks()) {
+      std::string track_id = track->id();
+      {
+        std::lock_guard<std::mutex> lock(local_stream_tracks_info_mutex_);
+        if (local_stream_tracks_info_.find(track_id) ==
+            local_stream_tracks_info_.end()) {
+          stream_track_info = std::make_pair(track_id, stream_id);
+          local_stream_tracks_info_.insert(stream_track_info);
+        }
+      }
+    }
+    for (const auto& track : media_stream->GetVideoTracks()) {
+      std::string track_id = track->id();
+      {
+        std::lock_guard<std::mutex> lock(local_stream_tracks_info_mutex_);
+        if (local_stream_tracks_info_.find(track_id) ==
+            local_stream_tracks_info_.end()) {
+          stream_track_info = std::make_pair(track_id, stream_id);
+          local_stream_tracks_info_.insert(stream_track_info);
+        }
+      }
+    }
+    if (on_success) {
+      on_success(stream);
+    }
```

Comment:
this is a semantic difference with `Publish()` where `on_success()` is called after the two vectors below are updated. lets move it to line 275 (just above `DrainPendingStreams()`) in case its important